### DISCUSSION TITLE
Adding a config option to enable firmware downgrades

### DISCUSSION THREAD

**Discussion by ast0815** [zigpy/zigpy#1293] [Ideas]
Hi all, the wiki has a section in it about downgrading firmwares: https://github.com/zigpy/zigpy/wiki/OTA-Device-Firmware-Updates#downgrading-your-devices-firmware

I am in the situation where I need to downgrade the firmware of a smart plug, but I am struggeling to follow those isntructions. Mostly I don't know how to edit that source file in my Home Assistant install (HomeAssitantOS on a RasPi). And in general it seems like a huge faff. Also the line number does no longer seem to be correct, but that is a minor issue.

So I wondered if it would be a good idea to add a config option that would just set the behaviour of the "should I update" check to be like the modified line suggested by the wiki. It would make downgrading a lot easier without having to dig too much in the innards of HA. As far as I can tell, it would only take adding the option and then putting an if statement there to check whether that option is set or not. The default behaviour would still stay the same and people would have to actively look for this option, so disclaimers can be put in the documentation to warn about firmware up/downgrade loops. We could also put in a check to make sure that this option is only used when the online updates are disabled, and only the local folder is used as a source of ota files.

If people think that this is an acceptable thing to do, I would even volunteer to prepare a PR for this (as soon as I find the time).

Do you think this is worthwhile? Or is there maybe already an easier way to downgrade firmwares with ZHA?

**Comment by TheJulianJES** [comment]
> I am in the situation where I need to downgrade the firmware of a smart plug

Which one? The Aqara EU smart plug?

**Reply by ast0815** [reply]
Yup. Old one works. Has been working for months. New one with newer firmware does not report power. Online searches suggest the new firmware might be to blame.

**Reply by TheJulianJES** [reply]
That plug checks if the "new firmware image" has a higher version number than the old one. A config option like this for zigpy would not work.
Check out these instructions here if you have the round Aqara EU plug: https://github.com/Koenkk/zigbee2mqtt/issues/13903#issuecomment-1244720072
Make sure to use the v99 file. 

Do note that ZHA's power measurement works perfectly with all firmware versions of the plug. However, the newer firmware versions of the plug require a somewhat up-to-date firmware on the coordinator. So, I'm guessing you're running really old firmware on your coordinator and should probably update anyway.
The downgrade of the plug is only really required if you have issues with it toggling on/off randomly when using IKEA remotes on older firmware versions on the same network.

**Reply by ast0815** [reply]
Oh, thanks a lot for that link.

Yeah, I also found the references telling me to update my coordinator firmware, but those also stated that in the worst case this could ruin my whole network, so I was hesitant to go that route. Aside from it looking like a not-so-userfriendly process to update the RaspBee II. I'll consider it.

**Comment by Hedda** [comment]
> the wiki has a section in it about downgrading firmwares: https://github.com/zigpy/zigpy/wiki/OTA-Device-Firmware-Updates#downgrading-your-devices-firmware

A tip is that can use the `ota_notify` feature in zha-toolkit for adding and updating OTA images without restarting Home Assistant -> https://github.com/mdeweerd/zha-toolkit#ota_notify---downloadtrigger-device-fw-update`

> I also found the references telling me to update my coordinator firmware, but those also stated that in the worst case this could ruin my whole network, so I was hesitant to go that route. Aside from it looking like a not-so-userfriendly process to update the RaspBee II. I'll consider it.

Normally I would suggest that a tip is to use [zigpy-cli](https://github.com/zigpy/zigpy-cli) as a external stand-alone command line tool for manual backup and restore on an other computer if ZHA's built-in backup features feels too much of a black box, however if you have ConBee or RaspBee with deconz firmware older than version 26720700 then there is a bug in it that prevents ZHA and zigpy from exporting backups, so it is a catch-22, if so then you will need to first upgrade deconz firmware to version 26720700 or newer before you can do export a backup. An alternative option could be to try to use deCONZ/Phoscon software application to perform backup and update firmware, but not sure if deCONZ/Phoscon software application allow you to perform backup if Zigbee network was already created in ZHA. See related info here -> https://skyconnect.home-assistant.io/migrate-conbee/

**Comment by Hedda** [comment]
> The default behaviour would still stay the same and people would have to actively look for this option, **so disclaimers can be put in the documentation to warn about firmware up/downgrade loops.** We could also put in a check to make sure that this option is only used when the online updates are disabled, and only the local folder is used as a source of ota files.

Oh, and since it is a wiki you can maybe edit and update it with that information yourself? 😉 

https://github.com/zigpy/zigpy/wiki/OTA-Device-Firmware-Updates#downgrading-your-devices-firmware

**Reply by ast0815** [reply]
That Wiki article already has a warning:

> Warning: This can lead to unwanted up/downgrades of devices, and may further lead to uncontrolled looping of up/downgrading your devices if you do not know what you are doing. ❗

I just meant that we can add this warning to the documentation of the proposed config option, if it ever gets implemented.

**Comment by ast0815** [comment]
Just to give people who find this thread some closure: I used the an sd-card imager from https://phoscon.de/de/conbee2/sdcard  with the instructions from https://github.com/dresden-elektronik/deconz-rest-plugin/wiki/Update-deCONZ-manually#update-in-raspbian to upgrade my RaspBee II firmware. Took a bit of trial and error, but it worked, and after reconnecting the Aqara plug I now get live power readings.

That removes at least _my_ need to downgrade any firmwares, so I will close this discussion.

**Comment by ast0815** [comment]
Hello, I am back to believing that I need to downgrade my Aqara plug firmware. Some of my plugs have firmware v41, which randomly switches the plug on and off if a sonoff button that connects directly to the plug is pressed. The older v32 firmware reportedly does not have that issue. See e.g. https://github.com/Koenkk/zigbee2mqtt/issues/13903

I am now trying again to look for a simple solution to downgrade the firmware. A simple switch in the zigpy configuration would be ideal. With as many explicit warnings as we like.

**Reply by TheJulianJES** [reply]
For this, you need firmware with a modified image version, since the plug only accepts (seemingly) newer ones. Use the v99 file from that thread for your plug. See my comments there for more info.

No need to modify or zigpy to allow for this downgrade, since it and the plug technically don't see it as one.
Just make sure to delete the file after installing it, so you won't try to install it again.

**Reply by puddly** [reply]
We actually "somewhat" support downgrades via a service call (https://github.com/zigpy/zha/blob/a117aa1c23b3413af1cc97076d840ba146180df0/zha/application/platforms/update.py#L266-L281) but we don't have any logic for forcing that specific image to be installed if the device expects the downgrade to have a higher version. Maybe we can do `self.device.device.update_firmware(..., force=(version is not None))` to force install any image that is listed as a downgrade?

**Reply by ast0815** [reply]
It worked well with the modded image, thanks!

But I still think it would be a good addition to be able to do that without having to modify some bytes in an obscure binary file. It's great if one can find such a file online, but what if not? I could maybe eventually figure out how to do this myself, but would prefer if I did not have to. Certainly not a high-priority request, though. :)
